### ISSUE TITLE
Question about Preprocessing

### ISSUE THREAD

**Issue by marcohuber** [IrvingMeng/MagFace#23]
Hello,
first of all thank you for providing the code, models and preprocessed images!

I tried to reproduce the results on LFW and with the preprocessed images you provided I also get the values you reported.
But when I preprocess the images from LFW by myself using the preprocessing from the ArcFace repository ([https://github.com/deepinsight/insightface](https://github.com/deepinsight/insightface)), the performance is a lot worse (~0.96) than with your images (~0.998).

I don't know if the authors of Arcface have changed the preprocessing code over time, but would it be possible to share the preprocessing script you used to create the images? Or if not, could you tell me what you changed to create the images?

Thanks in advance and thanks again for providing the code and models!
Best,
Marco

**Comment by IrvingMeng** [comment]
Hi Marco,

Thanks for trying MagFace. The test images we provided are extracted from InsightFace. The preprocessing step should be 
the same unless they change the code. 
Could you double-check whether five facial landmarks are correctly generated and alignment template is consistent with [here](https://github.com/deepinsight/insightface/blob/cdc3d4ed5de14712378f3d5a14249661e54a03ec/python-package/insightface/utils/face_align.py#L24-L26)?

Best,
Qiang

**Comment by marcohuber** [comment]
Hi Qiang,
so I checked it again. I extracted the five landmarks using the originally used MTCNN. It seems like InsightFace removed MTCNN and replaced it with RetinaFace. I think that landmark detection should not have such a big influence, but I still use MTCNN (as stated in the papers) and then transform it with these values: 

`
arcface_src = np.array(
    [[38.2946, 51.6963], [73.5318, 51.5014], [56.0252, 71.7366],
     [41.5493, 92.3655], [70.7299, 92.2041]],
    dtype=np.float32)`

If I now compare my results with the images you provided, I can see small differences, e.g.:
(Mine)
![Abel_Pacheco_0001](https://user-images.githubusercontent.com/38668954/133229371-cea69b53-024b-451a-9cef-855a8ea5de16.jpg)
(Yours)
![0](https://user-images.githubusercontent.com/38668954/133229462-5e5ef1e0-557d-47a7-bb24-f23a78b4f619.jpg)

which lead to a change in performance from 0.99 to 0.96 on the LFW benchmark. I also checked and used their original preprocessing from [2018](https://github.com/deepinsight/insightface/blob/60bb5829b1d76bfcec7930ce61c41dde26413279/src/common/face_preprocess.py) but it didn't get any better. My problem is that I would like to compare my approach with yours on other datasets as well, but I can't get the correct preprocessing. 

Did you use exactly the preprocessing you linked to above? They have at least made small changes now and then and it is also no longer MTCNN-based as originally in the ArcFace paper. It just confuses me because in your paper it is stated you use the same preprocessing as Arcface, but the ArcFace people don't bother to keep the preprocessing the same and changed it. Would it be possible for you to share your preprocessing script with me? If needed privately and under the assurance that I won't redistribute it, if it's a problem for your company otherwise.

Sorry for the inconvenience and thanks again,
Marco

**Comment by IrvingMeng** [comment]
Hi Marco,

The training/test datasets used are directly from InsightFace without any further modifications. It takes me a while to reproduce that process. The image below shows the result, where the folder 'data/' is from my repo while 'faces_emore/' is from InsightFace. You can see that the me5sum values are the same for the corresponding images. 

Here is how I reproduce the process:
1. download `faces_emore.zip` (MS1M-ArcFace (85K ids/5.8M images) [5,7]) from https://github.com/deepinsight/insightface/tree/master/recognition/_datasets_ 
2. unzip it to the folder `/face/irving/tmp/`
3. use https://github.com/wujiyang/Face_Pytorch/blob/master/utils/load_images_from_bin.py to extract the images from the `face_emore/lfw.bin`. Remember to comment out lines 64-65 while uncomment out lines 62,63, 66 before usage. Besides, I replace `str(idx+1).zfill(5)` to `str(idx)` in line 55 to align the filenames.

![screenshot-20210915-131149](https://user-images.githubusercontent.com/10399303/133374532-36400dd6-c128-4246-9f3d-c2abad682482.png)

Could you also evaluate your processed lfw with the InsightFace models to make sure that the testing protocols are correct? If that's not the problem, we can then check the processing with maintainers of InsightFace.

Best, 
Qiang Meng

**Comment by kormalev** [comment]
@marcohuber I would double-check the channel order (RGB vs BGR) and mean/std constants used to normalize input tensors (0.5/0.5 for InsightFace models vs 0.0/1.0 for MagFace).

**Comment by marcohuber** [comment]
Hi,
@IrvingMeng thank you for the effort! I reproduced your process and got the same images. For the testing protocols I mainly adopted your evaluation script and with the images preprocessed by InsightFace I achieve ~99,8%, but if I change the images to the images I preprocessed it drops to ~96,7%.
I just opend an [issue](https://github.com/deepinsight/insightface/issues/1753) at the InsightFace Github and I'll wait for their response. 
Thanks for your help and our efforts, I think this issue can be closed.

@kormalev I double-checked it and it seems to be correct. With the preprocessed image by InsightFace I am able to achieve the reported performance with my code, therefore the difference in alignment seems to be the only reason for the performance drop. I mean, it totally makes sense as the model was trained on a different alignment. But thanks for the idea!

Best,
Marco
